Fix problems on changing 'multiple' state SELECT element to 'single' state

Ahmad-S792 opened the pull request:
#### febbe13cd11cd845645ec472954d2048e3289cda
<pre>
Fix problems on changing &apos;multiple&apos; state SELECT element to &apos;single&apos; state

<a href="https://bugs.webkit.org/show_bug.cgi?id=252969">https://bugs.webkit.org/show_bug.cgi?id=252969</a>
rdar://problem/106264081

Reviewed by Aditya Keerthi.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - <a href="https://chromium.googlesource.com/chromium/src.git/+/ed6cd6b785c97d4a678c262bbf5e62bf46e5a2e8">https://chromium.googlesource.com/chromium/src.git/+/ed6cd6b785c97d4a678c262bbf5e62bf46e5a2e8</a>

This patch fixes two problems.

* Setting &apos;multiple&apos; IDL attribute to false and removing &apos;multiple&apos; content
  attribute had different behavior.
  They should be identical.  This patch removes HTMLSelectElement::setMultiple(), and
  add [Reflect] to &apos;multiple&apos; IDL attribute.

* If multiple SELECT without selected OPTIONs is changed to single, we should run
  reset().
  All other browsers agree with this behavior.

* Source/WebCore/html/HTMLSelectElement.cpp:
(HTMLSelectElement::setMultiple): Deleted
(HTMLSelectElement::parseMultipleAttribute): Update for &apos;reset&apos;
* Source/WebCore/html/HTMLSelectElement.h: Remove &apos;setMultiple&apos;
* Source/WebCore/html/HTMLSelectElement.idl: Update &apos;multiple&apos; to have [Reflect]
* Source/WebKitLegacy/mac/DOM/DOMHTMLSelectElement.mm: Update return instead of &apos;setMultiple&apos;
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMHTMLSelectElement.cpp:
(webkit_dom_html_select_element_set_multiple): Ditto
* LayoutTests/fast/forms/select-ask-for-reset.html: Sync up-to-date from Chromium Source
* LayoutTests/fast/forms/select-ask-for-reset-expected.txt: Rebaselined
* LayoutTests/fast/dom/HTMLSelectElement/change-multiple-preserve-selection.html: Rebaselined
* LayoutTests/fast/dom/HTMLSelectElement/change-multiple-preserve-selection-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/261380@main">https://commits.webkit.org/261380@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e611e5df4c59adc0ddc274c2363516c3be87e001

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111395 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20532 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/10 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3205 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120184 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11719 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99420 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98211 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/6 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44928 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13032 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/6 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86826 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13540 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18984 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51987 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7922 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15502 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKitLegacy/mac/DOM/DOMHTMLSelectElement.mm:
```diff
@@ -91,8 +91,7 @@ - (BOOL)multiple
 {
-    WebCore::JSMainThreadNullState state;
-    IMPL->setMultiple(newMultiple);
+    return;
```
I don't think this is safe to remove.

Ahmad-S792 replied:
Done.

pxlcoder commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -432,18 +432,6 @@ void HTMLSelectElement::optionElementChildrenChanged()
 
-void HTMLSelectElement::setMultiple(bool multiple)
-{
-    bool oldMultiple = this->multiple();
-    int oldSelectedIndex = selectedIndex();
-    setBooleanAttribute(multipleAttr, multiple);
-
-    // Restore selectedIndex after changing the multiple flag to preserve
-    // selection as single-line and multi-line has different defaults.
-    if (oldMultiple != this->multiple())
-        setSelectedIndex(oldSelectedIndex);
```
Since I don't think it's safe to remove the WebKitLegacy method, we should keep this method around.

Though it can just call `setBooleanAttribute(multipleAttr, multiple);` now.

nt1m replied:
Can't we just call `IMPL->setBooleanAttribute(multipleAttr, newMultiple);` directly from WebKitLegacy instead of keeping this around?

pxlcoder replied:
Good point! Let's do that.

Ahmad-S792 replied:
@nt1m - Thanks for guidance. Pushed changes in new commit.

pxlcoder commented on Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMHTMLSelectElement.cpp:
```diff
@@ -440,10 +440,7 @@ gboolean webkit_dom_html_select_element_get_multiple(WebKitDOMHTMLSelectElement*
-    WebCore::JSMainThreadNullState state;
-    g_return_if_fail(WEBKIT_DOM_IS_HTML_SELECT_ELEMENT(self));
-    WebCore::HTMLSelectElement* item = WebKit::core(self);
-    item->setMultiple(value);
```
We should also preserve the functionality of this method by calling `setBooleanAttribute`.

Ahmad-S792 replied:
`item->setBooleanAttribute(multipleAttr, newMultiple);`? Will do it once I confirm that the test cases don't need any further rebaselining. :-) Will aim to get it over the line tonight.

pxlcoder replied:
`item->setBooleanAttribute(WebCore::HTMLNames::multipleAttr, value);` in this case.

(matching the rest of the file / parameter name)

Ahmad-S792 replied:
I pushed wrong bit. :-( Fixing it now.

Ahmad-S792 replied:
Fixed in next push.

pxlcoder reviewed (approved):
r=me once all tests are passing.

Ahmad-S792 committed:
Fix problems on changing 'multiple' state SELECT element to 'single' state

https://bugs.webkit.org/show_bug.cgi?id=252969
rdar://problem/106264081

Reviewed by Aditya Keerthi.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - https://chromium.googlesource.com/chromium/src.git/+/ed6cd6b785c97d4a678c262bbf5e62bf46e5a2e8

This patch fixes two problems.

* Setting 'multiple' IDL attribute to false and removing 'multiple' content
  attribute had different behavior.
  They should be identical.  This patch removes HTMLSelectElement::setMultiple(), and
  add [Reflect] to 'multiple' IDL attribute.

* If multiple SELECT without selected OPTIONs is changed to single, we should run
  reset().
  All other browsers agree with this behavior.

* Source/WebCore/html/HTMLSelectElement.cpp:
(HTMLSelectElement::setMultiple): Deleted
(HTMLSelectElement::parseMultipleAttribute): Update for 'reset'
* Source/WebCore/html/HTMLSelectElement.h: Remove 'setMultiple'
* Source/WebCore/html/HTMLSelectElement.idl: Update 'multiple' to have [Reflect]
* Source/WebKitLegacy/mac/DOM/DOMHTMLSelectElement.mm: Update return instead of 'setMultiple'
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMHTMLSelectElement.cpp:
(webkit_dom_html_select_element_set_multiple): Ditto
* LayoutTests/fast/forms/select-ask-for-reset.html: Sync up-to-date from Chromium Source
* LayoutTests/fast/forms/select-ask-for-reset-expected.txt: Rebaselined
* LayoutTests/fast/dom/HTMLSelectElement/change-multiple-preserve-selection.html: Rebaselined
* LayoutTests/fast/dom/HTMLSelectElement/change-multiple-preserve-selection-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/261380@main
